src/current/v23.2/read-committed.md

```diff
@@ -0,0 +1,943 @@
+
+- [Constraint]({% link {{ page.version.version }}/constraints.md %}) violations will abort transactions at all isolation levels.
+
+- In rare cases, a `WriteTooOld` or `RETRY_SERIALIZABLE` error will be returned to the client if a statement has already begun streaming a partial result set back to the client and cannot retry transparently. By default, the result set is buffered up to 16 KiB before overflowing and being streamed to the client. You can configure the result buffer size using the [`sql.defaults.results_buffer.size`]({% link {{ page.version.version }}/cluster-settings.md %}#setting-sql-defaults-results-buffer-size) cluster setting or the [`results_buffer_size`]({% link {{ page.version.version }}/session-variables.md %}) session variable.
+
+### Concurrency anomalies
+
+Statements in concurrent `READ COMMITTED` transactions [can interleave with each other](#write-skew-tolerance). This can create concurrency anomalies that are not permitted under `SERIALIZABLE` isolation, which places concurrent transactions into a [serializable ordering]({% link {{ page.version.version }}/demo-serializable.md %}).
+
+{{site.data.alerts.callout_success}}
+The behavior described in this section assumes the use of non-locking reads. Concurrency anomalies can be prevented through the selective use of [locking reads](#locking-reads), which can also increase latency due to [lock contention]({% link {{ page.version.version }}/performance-best-practices-overview.md %}#transaction-contention).
+{{site.data.alerts.end}}
+
+#### Write skew anomaly
+
+Because a `READ COMMITTED` transaction's [read timestamp can skew from its write timestamp](#write-skew-tolerance), the following sequence of operations is possible: 
+
+1. Transaction `A` reads row `R` in table `T` at timestamp `1`.
+1. Transaction `B` writes to row `R` in table `T` and commits at timestamp `2`.
+1. Transaction `A` writes to table `T` and commits at timestamp `3`.
+
+Under `SERIALIZABLE` isolation, transaction `A` would have [refreshed the read]({% link {{ page.version.version }}/architecture/transaction-layer.md %}#read-refreshing) from timestamp `1` and found it to be incorrect at timestamp `3`. Instead of writing to table `T`, the transaction would have aborted with a [`RETRY_SERIALIZABLE`]({% link {{ page.version.version }}/transaction-retry-error-reference.md %}#retry_serializable) error, prompting the client to retry the transaction.
```

Comment:
The mention of "refreshing" here again seems like it may be too deep for the docs on Read Committed. It's an implementation detail of how serializable transactions in CRDB provide serializable.

I say this while also acknowledging that finding the right balance is going to be really hard! Readers of our docs come equipped with vastly different backgrounds and will learn best from different information. Talking about implementation techniques will be very effective for database experts. However, it might make an already difficult topic entirely unapproachable for less experienced database users. It almost feels like we need two docs coming at the topic from two different directions.

I sincerely believe that writing the docs for Read Committed will be the hardest part of the project. Database isolation levels are confusing in large part because they began as a collection of implementation alternatives that were later retrofitted with some attempt to have understandable behavior. We've done our best to make Cockroach's implementation easier to understand than most (e.g. it's based on per-statement read snapshots, not race conditions between scanning and locking of different indexes), but it's still tricky to teach.

During our early research, we were surprised to find that Oracle actually also has one of the more reasonable implementation of isolation levels, with an approach of per-statement vs. per-txn read snapshots like ours. It might be worth looking at how they document this (keeping in mind that Oracle calls Snapshot isolation "Serializable", so all isolation levels are susceptible to write skew): https://docs.oracle.com/cd/E25054_01/server.1111/e25789/consist.htm

Reply:
This feedback makes a lot of sense. I suddenly realize that the `demo-serializable.md` doc that I've been linking here as the analogous Serializable doc is a much-streamlined tutorial on transaction behavior, and our existing `transaction-layer.md` architecture doc is what delves deeper into the implementation details. I don't intend to get into the architecture docs with this PR, so I tried to strike a middle ground here between implementation and practical details. But I see that synthesizing both approaches into one doc is distracting from its practical value. I'll take another pass and deflate the implementation details, which we could reuse for a future update to the architecture doc.